### ISSUE TITLE
AssertionError when using predefScript.sc

### ISSUE THREAD

**Issue by francisdb** [com-lihaoyi/mill#820]
Installed mill with homebrew

```
> mill --version
Mill Build Tool version 0.6.1
Java version: 13.0.2, vendor: N/A, runtime: /usr/local/Cellar/openjdk/13.0.2+8_2/libexec/openjdk.jdk/Contents/Home
Default locale: en_BE, platform encoding: UTF-8
OS name: "Mac OS X", version: 10.15.3, arch: x86_64
```
Followed docs here
http://www.lihaoyi.com/mill/page/configuring-mill.html#global-configuration

And when I start mill I see this:
```
> mill api:compile
Compiling /Users/myuser/.mill/ammonite/predefScript.sc
java.lang.AssertionError: assertion failed
  scala.Predef$.assert(Predef.scala:208)
  mill.define.Segment$Label.<init>(Ctx.scala:14)
  mill.define.Segment$Label$.apply(Ctx.scala:13)
  mill.define.Segment$Label$.apply(Ctx.scala:13)
  scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
  scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
  scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
  scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:39)
  scala.collection.TraversableLike.map(TraversableLike.scala:237)
  scala.collection.TraversableLike.map$(TraversableLike.scala:230)
  scala.collection.AbstractTraversable.map(Traversable.scala:108)
  mill.define.Segments$.labels(Ctx.scala:57)
  ammonite.predef.predefScript$.<init>(predefScript.sc:54)
  ammonite.predef.predefScript$.<clinit>(predefScript.sc)
```

Assertion errors are not very user friendly, this could give some more information on what is failing and why

https://github.com/lihaoyi/mill/blob/85aa6bdb1bd1d08fb2739a4fd75dd76b0ab6652c/main/core/src/define/Ctx.scala#L14

**Comment by hypatian** [comment]
I'm having this same problem, both with the homebrew version (which appears to just be the bootstrap script) and with the bootstrap script by hand. I added to the assert to see where the `.` is being found, and discovered that the complaint is about the `.` in `.mill`. I'll see if I can figure out how to make this stop failing.

**Comment by lefou** [comment]
> > mill api:compile

What is this line supposed to do? If `api` is the module and `compile` the target, the correct invocation is: `mill api.compile`.

**Comment by francisdb** [comment]
> > > mill api:compile
> 
> What is this line supposed to do? If `api` is the module and `compile` the target, the correct invocation is: `mill api.compile`.

Let me turn this question around.
Why did mill not tell me I did something the wrong way and point me to the correct way of doing things?

Anyway, same error with `mill api.compile`

**Comment by hypatian** [comment]
It looks like this problem has been around since mill 0.5.5. I have a fix and am putting in a pull request for it now. Not sure how to add a test for it, I'll look into that after.

**Comment by ml86** [comment]
This issue is still present in mill version 0.7.4:
```
> mill -version
Mill Build Tool version 0.7.4
Java version: 1.8.0_265, vendor: Private Build, runtime: /usr/lib/jvm/java-8-openjdk-amd64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "Linux", version: 4.15.0-118-generic, arch: amd64
```
```
> mill _.compile
Compiling /home/ml/.mill/ammonite/predefScript.sc
java.lang.AssertionError: assertion failed
  scala.Predef$.assert(Predef.scala:267)
  mill.define.Segment$Label.<init>(Ctx.scala:14)
  mill.define.Segment$Label$.apply(Ctx.scala:13)
  mill.define.Segment$Label$.apply(Ctx.scala:13)
  scala.collection.immutable.ArraySeq.$anonfun$map$1(ArraySeq.scala:71)
  scala.collection.immutable.ArraySeq.$anonfun$map$1$adapted(ArraySeq.scala:71)
  scala.collection.immutable.ArraySeq$.tabulate(ArraySeq.scala:192)
  scala.collection.immutable.ArraySeq$.tabulate(ArraySeq.scala:171)
  scala.collection.ClassTagIterableFactory$AnyIterableDelegate.tabulate(Factory.scala:664)
  scala.collection.immutable.ArraySeq.map(ArraySeq.scala:71)
  scala.collection.immutable.ArraySeq.map(ArraySeq.scala:35)
  mill.define.Segments$.labels(Ctx.scala:57)
  ammonite.predef.predefScript$.<init>(predefScript.sc:37)
  ammonite.predef.predefScript$.<clinit>(predefScript.sc:36)
```

After a quick look in the code I guess that assertion is there in order to assure that version schemas can be parsed? If so, how about allowing leading dots in segments?

**Comment by lefou** [comment]
I think we should change the `assert` into a `require`, and should also add a motivation for it. I had a very glance at it, and it looks to me the assert/require is related to fact that labels must meet the naming restrictions of package and object names. So the only chance to support dots might be to find a proper encoding.

**Comment by DamianReeves** [comment]
Any updates on this, it seems to still be around in 0.9.8?

**Comment by lefou** [comment]
The issue is the dot in the path name, so a work around is to use a directory without a dot in its path.

**Comment by lefou** [comment]
TBH, I tackled this issue multiple times, but I failed to find any evidence or test that gives some clue which cases we try to avoid by asserting that labels do not contain a dot. I only can guess. What's I can clearly see, once we allow dot's in labels, we can no longer parse targets on the commandline. Currently, we split the target names at each dot. As the real issue are hidden files/dirs, which start with a dot in the name (on linux/mac) we could try to just allow this and handle it specially. Parsing such cases is possible. Whoever want's to do it, starting with a test case is probably the best way to get going.

**Comment by DamianReeves** [comment]
As an alternative is there a way we could perhaps change the "HOME" directory for mill, meaning can I force it not to look for these in `~/.mill/ammonite/` and perhaps to look at a custom location which does not include the dot?

**Comment by lefou** [comment]
You can try to specify an alternative ammonite home directory with `mill --home /path/to/home`. This is by default: `$HOME/.mill/ammonite`.

**Comment by DamianReeves** [comment]
I always felt it would be nice to be able to have local predefs, that users could perhaps exclude as part of their project loading.

What would I need to add to my `build.sc` to have it load scripts that execute before the rest of the build.
I'm honestly setting proxy settings that I don't want to check into the repo or need to remember to pass command line arguments or set environment variables.

My thought is if I could do something like:

```
 - build.sc
 - build/
    - config/
        - user/
            - predefScript.sc  
```

`.gitignore`

```
build/config/user/
```

I get that loaded as the first thing in my build, but only if it exists (i.e. I don't want it to be required).

**Comment by DamianReeves** [comment]
> You can try to specify an alternative ammonite home directory with `mill --home /path/to/home`. This is by default: `$HOME/.mill/ammonite`.

So this works, but is it possible to change home using an environment variable?

**Comment by lefou** [comment]
> > You can try to specify an alternative ammonite home directory with `mill --home /path/to/home`. This is by default: `$HOME/.mill/ammonite`.
> 
> So this works, but is it possible to change home using an environment variable?

I don't think so.

**Comment by DamianReeves** [comment]
> I don't think so.

So on Mac the homebrew install of mill leaves us with the expectation of mill being in `~/.mill` this is problematic when like me you have a Mac behind a corporate proxy or with custom repositories which need to be set ahead of time.

An environment variable for `MILL_HOME` would solve so much.

**Comment by lefou** [comment]
> An environment variable for `MILL_HOME` would solve so much.

Hm, I have no problem with using an environment variable. But I think `MILL_HOME` isn't the best name. It's actually the internal cache dir used by Ammonite. And a good default should probably be a directory under `$XDG_CACHE_DIR`, as we already do with the wrapper script.

**Comment by DamianReeves** [comment]
So coursier uses `directories-jvm` to arrive at the proper location to store configuration files on a per OS basis. Perhaps mill/ammonite should also use such logic. This results in coursier looking for global properties in expected OS directories as outlined here: https://scalameta.org/metals/docs/troubleshooting/proxy/#via-global-properties-file.

**Comment by lefou** [comment]
This issue now has a fix in PR #1637. Any feedback appreciated.
